### PR TITLE
DataFrameClustering module

### PR DESCRIPTION
**Pull request by r-englund**
Added module/processor for clustering of data in dataframes, supports Agglomerative, Spectral and KMeans clustering using SikitLearn and SciPy. 

TODO before ready for review
- [x] Check/Add example workspaces
- [x] Check documentation (Code / Processors doc)
- [x] Read me

### CODE REVIEW

File: misc/dataframeclustering/src/processors/dataframeclustering.cpp [hunk 1]
```diff
@@ -0,0 +1,181 @@
...
+                {"complete", "complete"},
+                {"average", "average"},
+                {"weighted", "weighted"},
+                {"centroid", "centroid"},
+                {"median", "median"},
+                {"ward", "ward"}})
+
+    , columns_("columns", "Columns")
+    , script_(app->getModuleByType<DataFrameClusteringModule>()->getPath(ModulePath::Scripts) +
+              "/dataframeclustering.py") {
```
Comment (martinfalk) [inline misc/dataframeclustering/src/processors/dataframeclustering.cpp:88]: script missing in PR

File: misc/dataframeclustering/CMakeLists.txt [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+    include/inviwo/dataframeclustering/dataframeclusteringmoduledefine.h
+    include/inviwo/dataframeclustering/processors/dataframeclustering.h
+)
+ivw_group("Header Files" ${HEADER_FILES})
+
+set(SOURCE_FILES
+    src/dataframeclusteringmodule.cpp
+    src/processors/dataframeclustering.cpp
+)
+ivw_group("Source Files" ${SOURCE_FILES})
```
Comment (martinfalk) [inline misc/dataframeclustering/CMakeLists.txt:14]: `/dataframeclustering.py` should be added somewhere for packaging

File: misc/dataframeclustering/readme.md [hunk 3]
```diff
@@ -0,0 +1,3 @@
+# DataFrameClustering Module
+
+Description of the DataFrameClustering module
```
Comment (martinfalk) [inline misc/dataframeclustering/readme.md:3]: Suggestion:
This module provides the functionality for clustering the rows of a DataFrame. Supported clustering methods are _k-means_, _DBSCAN_, _agglomerative_, and _spectral_ clustering.
 The clustering is performed in python using the following modules: `...`.

File: misc/dataframeclustering/src/processors/dataframeclustering.cpp [hunk 4]
```diff
@@ -0,0 +1,181 @@
...
+
+namespace inviwo {
+
+// The Class Identifier has to be globally unique. Use a reverse DNS naming
+// scheme
+const ProcessorInfo DataFrameClustering::processorInfo_{
+    "org.inviwo.DataFrameClustering",  // Class identifier
+    "Data Frame Clustering",           // Display name
+    "Clustering",                      // Category
+    CodeState::Experimental,           // Code state
```
Comment (martinfalk) [inline misc/dataframeclustering/src/processors/dataframeclustering.cpp:42]: might change to stable

**Review by martinfalk — CHANGES_REQUESTED** [review]

File: misc/dataframeclustering/src/processors/dataframeclustering.cpp (hunk 1 above)
Comment (r-englund) [reply misc/dataframeclustering/src/processors/dataframeclustering.cpp:88]: Wiered, I copy-pasted the whole folder and did 'git add . '

**Comment by petersteneteg** [comment]
also enable the module in jenkins maybe?

**Comment by petersteneteg** [comment]
was the log renderer not in a different PR?

File: misc/dataframeclustering/data/scripts/dataframeclustering.py [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+    label, count = np.unique(labels,return_counts=True)
+    freq = list(zip(label,count))
+    firstPostive = np.argmax(label > -1)
+    
+    for i,(l,f) in enumerate(sorted(freq[firstPostive:], key=lambda x: x[1], reverse=True)):
+        res[labels == l] = i
+    return res
+        
+
+df = np.zeros([dataframe.rows,len(activeColumnHeaders)])
```
Comment (martinfalk) [inline misc/dataframeclustering/data/scripts/dataframeclustering.py:24]: are the global variables `dataframe`, `activeColumnHeaders`, and `method` provided by the environment? Maybe add comment at the top

File: misc/devtools/CMakeLists.txt [hunk 6]
```diff
@@ -16,6 +17,7 @@ ivw_group("Header Files" ${HEADER_FILES})
 set(SOURCE_FILES
     src/devtoolsmodule.cpp
     src/processors/eventlogger.cpp
+    src/processors/logrendererprocessors.cpp
```
Comment (martinfalk) [inline misc/devtools/CMakeLists.txt:20]: add `data/scripts/dataframeclustering.py`

File: misc/devtools/depends.cmake [hunk 7]
```diff
@@ -3,6 +3,7 @@
 set(dependencies
     #InviwoOpenGLModule # Example dependency 
     InviwoBrushingAndLinkingModule
+    InviwoFontRenderingModule
```
Comment (martinfalk) [inline misc/devtools/depends.cmake:6]: needed?

**Review by martinfalk — APPROVED** [review]

File: misc/dataframeclustering/CMakeLists.txt (hunk 2 above)
Comment (martinfalk) [reply misc/dataframeclustering/CMakeLists.txt:14]: automatically added due to location in `data/scripts/`

**Pull request merged**